example-crates/example_rpc/src/main.rs

```diff
@@ -0,0 +1,308 @@
+
+    let (args, keymap, index, db, init_changeset) =
+        example_cli::init::<RpcCommands, ChangeSet>(DB_MAGIC, DB_PATH)?;
+
+    let graph = Mutex::new({
+        let mut graph = IndexedTxGraph::new(index);
+        graph.apply_additions(init_changeset.indexed_additions);
+        graph
+    });
+
+    let chain = Mutex::new(LocalChain::from_changeset(init_changeset.chain_changeset));
+
+    let rpc_cmd = match args.command {
+        example_cli::Commands::ChainSpecific(rpc_cmd) => rpc_cmd,
+        general_cmd => {
+            let res = example_cli::handle_commands(
+                &graph,
+                &db,
+                &chain,
+                &keymap,
+                args.network,
+                |_| Err(anyhow::anyhow!("use `tx` instead")),
+                general_cmd,
+            );
+            db.lock().unwrap().commit()?;
+            return res;
+        }
+    };
+
+    let rpc_client = {
+        let a = rpc_cmd.rpc_args();
+        Client::new(
+            &a.url,
+            match (&a.rpc_cookie, &a.rpc_user, &a.rpc_password) {
+                (None, None, None) => Auth::None,
+                (Some(path), _, _) => Auth::CookieFile(path.clone()),
+                (_, Some(user), Some(pass)) => Auth::UserPass(user.clone(), pass.clone()),
+                (_, Some(_), None) => panic!("rpc auth: missing rpc_pass"),
+                (_, None, Some(_)) => panic!("rpc auth: missing rpc_user"),
+            },
+        )?
+    };
+
+    match rpc_cmd {
+        RpcCommands::Scan {
+            fallback_height,
+            lookahead,
+            live,
+            ..
+        } => {
+            graph.lock().unwrap().index.set_lookahead_for_all(lookahead);
+
+            let (chan, recv) = sync_channel::<(EmittedUpdate, u32)>(CHANNEL_BOUND);
+            let prev_cp = chain.lock().unwrap().tip();
+
+            let join_handle = std::thread::spawn(move || -> anyhow::Result<()> {
+                let mut tip_height = Option::<u32>::None;
+
+                for item in Emitter::new(&rpc_client, fallback_height, prev_cp).into_iter() {
+                    let item = item?;
+                    let is_block = !item.is_mempool();
+                    let is_mempool = item.is_mempool();
+
+                    if tip_height.is_none() || !is_block {
+                        tip_height = Some(rpc_client.get_block_count()? as u32);
+                    }
+                    chan.send((item, tip_height.expect("must have tip height")))?;
+
+                    if sigterm_flag.load(Ordering::Acquire) {
+                        return Ok(());
+                    }
+                    if is_mempool {
+                        if !live {
+                            return Ok(());
+                        }
+                        if await_flag(&sigterm_flag, LIVE_POLL_DUR_SECS) {
+                            return Ok(());
+                        }
+                    }
+                }
+                unreachable!()
```

Comment:
This `unreachable` thing biases me against providing this iterator API. It ended up conceptually weird to me. IWouldn't it be clearer just to emit blocks and process them in a `loop`.

Reply:
The way we are using it here is a bit weird. We have a flag for both "live" mode and "sync-once" mode. In my opinion, the iterator will be useful in a "sync-once" mode. The "unreachable" thing is not a necessity or a direct result of using the iterator.